tracer/tracer.go

```diff
@@ -0,0 +1,166 @@
+package tracer
+
+// TODO: change package
+
+import (
+	"time"
+
+	opentracing "github.com/opentracing/opentracing-go"
+)
+
+// Tracer extends the opentracing.Tracer interface with methods to
+// probe implementation state, for use by tracer consumers.
+type Tracer interface {
+	opentracing.Tracer
+	Options() Options
+}
+
+type SpanRecorder interface {
+	RecordSpan(span RawSpan)
+}
+
+type Options struct {
+	ShouldSample   func(traceID string) bool
+	Recorder       SpanRecorder
+	EnableSpanPool bool
+}
+
+// TODO: review ShouldSample
+func DefaultOptions() Options {
+	return Options{
+		ShouldSample: func(traceID string) bool { return true }, //{ return traceID%64 == 0 },
+	}
+}
+
+// NewWithOptions creates a customized Tracer.
+func NewWithOptions(opts Options) opentracing.Tracer {
+	rval := &tracerImpl{options: opts}
+	rval.textPropagator = &textMapPropagator{rval}
+	rval.binaryPropagator = &binaryPropagator{rval}
+	rval.accessorPropagator = &accessorPropagator{rval}
+	return rval
+}
+
+// New creates and returns a standard Tracer which defers completed Spans to
+// `recorder`.
+// Spans created by this Tracer support the ext.SamplingPriority tag: Setting
+// ext.SamplingPriority causes the Span to be Sampled from that point on.
+func New(recorder SpanRecorder) opentracing.Tracer {
+	opts := DefaultOptions()
+	opts.Recorder = recorder
+	return NewWithOptions(opts)
+}
+
+// Implements the `Tracer` interface.
+type tracerImpl struct {
+	options            Options
+	textPropagator     *textMapPropagator
+	binaryPropagator   *binaryPropagator
+	accessorPropagator *accessorPropagator
+}
+
+func (t *tracerImpl) StartSpan(operationName string, opts ...opentracing.StartSpanOption) opentracing.Span {
+	sso := opentracing.StartSpanOptions{}
+	for _, o := range opts {
+		o.Apply(&sso)
+	}
+	return t.StartSpanWithOptions(operationName, sso)
+}
+
+func (t *tracerImpl) getSpan() *spanImpl {
+	if t.options.EnableSpanPool {
+		sp := spanPool.Get().(*spanImpl)
+		sp.reset()
+		return sp
+	}
+	return &spanImpl{}
+}
+
+func (t *tracerImpl) StartSpanWithOptions(operationName string, opts opentracing.StartSpanOptions) opentracing.Span {
+	// Start time.
+	startTime := opts.StartTime
+	if startTime.IsZero() {
+		startTime = time.Now()
+	}
+
+	// Tags.
+	tags := opts.Tags
+
+	// Build the new span. This is the only allocation: We'll return this as
+	// an opentracing.Span.
+	sp := t.getSpan()
+	// Look for a parent in the list of References.
+	//
+	// TODO: would be nice if tracer did something with all
+	// References, not just the first one.
+ReferencesLoop:
+	for _, ref := range opts.References {
+		switch ref.Type {
+		case opentracing.ChildOfRef,
+			opentracing.FollowsFromRef:
+
+			refCtx := ref.ReferencedContext.(SpanContext)
```

Comment:
Note: In all the other Wavefront SDKs, we look to choose the first `childOf` reference as the parent, and fall back to `followsFrom` only if none are found. 

For example: https://github.com/wavefrontHQ/wavefront-opentracing-sdk-java/blob/dc58a758487b3139a3999b02e55edb08238f1143/src/main/java/com/wavefront/opentracing/WavefrontSpanBuilder.java#L155